examples/tools/scene_viewer.rs

```diff
@@ -0,0 +1,407 @@
+        projection_adjustment.y -= SCALE_STEP;
+    } else if key_input.just_pressed(KeyCode::Key8) {
+        projection_adjustment.y += SCALE_STEP;
+    } else if key_input.just_pressed(KeyCode::Key9) {
+        projection_adjustment.z -= SCALE_STEP;
+    } else if key_input.just_pressed(KeyCode::Key0) {
+        projection_adjustment.z += SCALE_STEP;
+    }
+    for (_, mut light) in query.iter_mut() {
+        light.shadow_projection.left *= projection_adjustment.x;
+        light.shadow_projection.right *= projection_adjustment.x;
+        light.shadow_projection.bottom *= projection_adjustment.y;
+        light.shadow_projection.top *= projection_adjustment.y;
+        light.shadow_projection.near *= projection_adjustment.z;
+        light.shadow_projection.far *= projection_adjustment.z;
+        if key_input.just_pressed(KeyCode::U) {
+            light.shadows_enabled = !light.shadows_enabled;
+        }
+    }
+
+    if key_input.just_pressed(KeyCode::L) {
+        *animate_directional_light = !*animate_directional_light;
+    }
+    if *animate_directional_light {
+        for (mut transform, _) in query.iter_mut() {
+            transform.rotation = Quat::from_euler(
+                EulerRot::ZYX,
+                0.0,
+                time.seconds_since_startup() as f32 * std::f32::consts::TAU / 30.0,
+                -std::f32::consts::FRAC_PI_4,
+            );
+        }
+    }
+}
+
+#[derive(Component)]
+struct CameraController {
+    pub enabled: bool,
+    pub initialized: bool,
+    pub sensitivity: f32,
+    pub key_forward: KeyCode,
+    pub key_back: KeyCode,
+    pub key_left: KeyCode,
+    pub key_right: KeyCode,
+    pub key_up: KeyCode,
+    pub key_down: KeyCode,
+    pub key_run: KeyCode,
+    pub walk_speed: f32,
+    pub run_speed: f32,
+    pub friction: f32,
+    pub pitch: f32,
+    pub yaw: f32,
+    pub velocity: Vec3,
+}
+
+impl Default for CameraController {
+    fn default() -> Self {
+        Self {
+            enabled: true,
+            initialized: false,
+            sensitivity: 0.5,
+            key_forward: KeyCode::W,
+            key_back: KeyCode::S,
+            key_left: KeyCode::A,
+            key_right: KeyCode::D,
+            key_up: KeyCode::E,
+            key_down: KeyCode::Q,
+            key_run: KeyCode::LShift,
+            walk_speed: 5.0,
+            run_speed: 15.0,
+            friction: 0.5,
+            pitch: 0.0,
+            yaw: 0.0,
+            velocity: Vec3::ZERO,
+        }
+    }
+}
+
+fn camera_controller(
+    time: Res<Time>,
+    mut mouse_events: EventReader<MouseMotion>,
+    key_input: Res<Input<KeyCode>>,
+    mut query: Query<(&mut Transform, &mut CameraController), With<Camera>>,
+) {
+    let dt = time.delta_seconds();
+
+    // Handle mouse input
+    let mut mouse_delta = Vec2::ZERO;
+    for mouse_event in mouse_events.iter() {
+        mouse_delta += mouse_event.delta;
+    }
+
+    for (mut transform, mut options) in query.iter_mut() {
```

Comment:
Can't this just be a query.single?

Reply:
Yes, you're right. Because the `CameraController` component is only added to one entity. Done.

Reply:
Though note that this code is copied and pasted from the shadow_biases example. It would probably be better if it were shared.

Reply:
Changed to `.get_single_mut()` as there are no entities with a `CameraController` while the scene is loading.